src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,1154 @@
+namespace ts.refactor.extractMethod {
+    const extractMethod: Refactor = {
+        name: "Extract Method",
+        description: Diagnostics.Extract_function.message,
+        getAvailableActions,
+        getEditsForAction,
+    };
+
+    registerRefactor(extractMethod);
+
+    /** Compute the associated code actions */
+    function getAvailableActions(context: RefactorContext): ApplicableRefactorInfo[] | undefined {
+        const rangeToExtract = getRangeToExtract(context.file, { start: context.startPosition, length: context.endPosition - context.startPosition });
+
+        const targetRange: TargetRange = rangeToExtract.targetRange;
+        if (targetRange === undefined) {
+            return undefined;
+        }
+
+        const extractions = extractRange(targetRange, context);
+        if (extractions === undefined) {
+            // No extractions possible
+            return undefined;
+        }
+
+        const actions: RefactorActionInfo[] = [];
+        const usedNames: Map<boolean> = createMap();
+
+        let i = 0;
+        for (const extr of extractions) {
+            // Skip these since we don't have a way to report errors yet
+            if (extr.errors && extr.errors.length) {
+                continue;
+            }
+
+            // Don't issue refactorings with duplicated names
+            const description = formatStringFromArgs(Diagnostics.Extract_function_into_0.message, [extr.scopeDescription]);
+            if (!usedNames.get(description)) {
+                usedNames.set(description, true);
+                actions.push({
+                    description,
+                    name: `scope_${i}`
+                });
+            }
+            i++;
+        }
+
+        if (actions.length === 0) {
+            return undefined;
+        }
+
+        return [{
+            name: extractMethod.name,
+            description: extractMethod.description,
+            inlineable: true,
+            actions
+        }];
+    }
+
+    function getEditsForAction(context: RefactorContext, actionName: string): RefactorEditInfo | undefined {
+        const length = context.endPosition === undefined ? 0 : context.endPosition - context.startPosition;
+        const rangeToExtract = getRangeToExtract(context.file, { start: context.startPosition, length });
+        const targetRange: TargetRange = rangeToExtract.targetRange;
+
+        const parsedIndexMatch = /scope_(\d+)/.exec(actionName);
+        if (!parsedIndexMatch) {
+            throw new Error("Expected to match the regexp");
+        }
+        const index = +parsedIndexMatch[1];
+        if (!isFinite(index)) {
+            throw new Error("Expected to parse a number from the scope index");
+        }
+
+        const extractions = extractRange(targetRange, context, index);
+        if (extractions === undefined) {
+            // Scope is no longer valid from when the user issued the refactor
+            // return undefined;
+            return undefined;

```

Comment:
This should be very rare, right?  Cases like the file refreshing in the background?